src/realm/query_engine.hpp

```diff
@@ -1744,27 +1744,49 @@ class StringNode : public StringNodeBase {
         TConditionFunction cond;
 
         for (size_t s = start; s < end; ++s) {
-            if constexpr (std::is_same_v<TConditionFunction, NotEqual>) {
+
+            // special handling for !=, <, <= , >, >= if the leaf is compressed and we have got a compressed string
+            // id.
+            if constexpr (realm::is_any_v<TConditionFunction, NotEqual, Greater, Less, GreaterEqual, LessEqual>) {
+
+                auto cmp = [](auto v) {
+                    if constexpr (std::is_same_v<TConditionFunction, NotEqual>) {
+                        return v != 0;
+                    }
+                    if constexpr (std::is_same_v<TConditionFunction, Less>) {
+                        return v < 0;
+                    }
+                    if constexpr (std::is_same_v<TConditionFunction, LessEqual>) {
+                        return v <= 0;
+                    }
+                    if constexpr (std::is_same_v<TConditionFunction, Greater>) {
+                        return v > 0;
+                    }
+                    if constexpr (std::is_same_v<TConditionFunction, GreaterEqual>) {
+                        return v >= 0;
+                    }
+                    REALM_UNREACHABLE();
+                };
+
                 if (m_leaf->is_compressed()) {
                     if (m_interned_string_id) {
-                        // The search string has been interned, so there might be a match
-                        // We can compare the string IDs directly
                         const auto id = m_leaf->get_string_id(s);
-                        if (m_string_interner->compare(*m_interned_string_id, *id) == 0) {
-                            // The value matched, so we continue to the next value
+                        if (cmp(m_string_interner->compare(*id, *m_interned_string_id)))
+                            return s;
+                        else
                             continue;
-                        }
                     }
-                    return s;
                 }
             }
+
             StringData t = get_string(s);
             if constexpr (case_sensitive_comparison) {
                 // case insensitive not implemented for: >, >=, <, <=
                 if (cond(t, m_string_value))
                     return s;
             }
             else {
+                StringData t = get_string(s);
```

Comment:
why is this necessary?

Reply:
for operators like `BeginsWith` and `EndsWith`

Reply:
The same line of code exists 7 lines up, so it seems redundant?